---
intitle: "class11"
author: "Phoebe LI"
date: "2/25/2022"
output: pdf_document
---

## 1. Bioconductor and DESeq2 setup
```{r}
library(BiocManager)
library(DESeq2)
```

## 2. Import countData and colData
```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

```{r}
head(counts)
```

```{r}
head(counts)
View(metadata)
```

```{r}
dim(counts)
```


> Q1 How many genes are in this dataset? 
 38694 genes
Q2. How many ‘control’ cell lines do we have? 
  4 controls
  
  
##  Toy differential gene expression
  
```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.counts <- counts[ ,control$id]
control.mean <- rowSums( control.counts )/4 
head(control.mean)

```

> Q3. How would you make the above code in either approach more robust?

```{r}
library(dplyr)
control <- metadata %>% filter(dex=="control")
control.counts <- counts %>% select(control$id) 
control.mean <- rowSums(control.counts)/4
head(control.mean)

```


> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

# Double check with the id
# can use all function here. all(c(T,T,T,F, T))
# how to check if anything false? you can use ! it will flip all the things
```{r}
metadata$id == colnames(counts)
colnames(counts)

```


```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/4 
names(treated.mean) <- counts$ensgene
```


# We will combine our meancount data for bookkeeping purposes.
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```


```{r}
colSums(meancounts ) 
```


> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

```{r}
plot(meancounts[,1],meancounts[,2], xlab="Control", ylab="Treated")
```



> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot? 
 geom_point
 
> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

```{r}

plot(meancounts[,1],meancounts[,2], log="xy", xlab="Control", ylab="Treated")
```

# We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```


```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The arr.ind=TRUE argument will clause which() to return both the row and column indices (i.e. positions) where there are TRUE values. In this case this will tell us which genes (rows) and samples (columns) have zero counts. We are going to ignore any genes that have zero counts in any sample so we just focus on the row answer. Calling unique() will ensure we dont count any row twice if it has zer entries in both samples. 

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

>Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level? 
 250
Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?
367
Q10. Do you trust these results? Why or why not?
No, all our analysis has been done based on fold change. We have not done anything yet to determine whether the differences we are seeing are significant. 


## 4. DESeq2 analysis
```{r}
library(DESeq2)
citation("DESeq2")
```


```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

# DESeq analysis

```{r}
dds <- DESeq(dds)
results(dds)
```

# Getting results
```{r}
res <- results(dds)
res
```


We can summarize some basic tallies using the summary function.
```{r}
summary(res)
res05 <- results(dds, alpha=0.05)
summary(res05)
```


# add annotation  Data
Let's install 
-BiocManager::install("AnnotationDbi")
-BiocManager::install("org.Hs.eg.db")
```{r}
# add annotation  Data
library("AnnotationDbi")
library("org.Hs.eg.db")
```

We can now use te 'mapIDs' function from the annotation function from the AnnotationDbi package to find 

```{r}
# First, let us see what is available in the annotation
columns(org.Hs.eg.db)
```


```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

head(res)
```



```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])
```


```{r}
write.csv(res[ord,], "deseq_results.csv")
```


```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)

```



# just to understand name function
```{r}
x<-c(1,5,10)
names(x)<-c("A","B","C")
x

```


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```


```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```




```{r}
attributes(keggres)
```

We can find out what is in this thing by calling the 'attributes' function
```{r}
head(keggres$less, 3)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```


# save my result
```{r}
write.csv(res[ord,], "deseq_results.csv")
```


```{r}
sessionInfo()
```

